### PR TITLE
detect: Add byte_math detector

### PR DESCRIPTION
**Pull request by jlucovsky**
Continuation of #4805 

This PR adds `byte_math` to the suite of `byte` functions with a strict implementation based on the Snort documents.

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: [2311](https://redmine.openinfosecfoundation.org/issues/2311)

Describe changes:
- Address review comments

### CODE REVIEW

File: src/detect-bytemath.c [hunk 1]
```diff
@@ -0,0 +1,1284 @@
...
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file
+ *
+ * \author Jeff Lucovsky <jeff@lucovsky.org>
+ */
+
+/*
```
Comment (victorjulien) [inline src/detect-bytemath.c:24]: can you remove this and just other than the reference to "Section 3.5.34, Snort Manual"

File: src/detect-bytemath.c [hunk 2]
```diff
@@ -0,0 +1,1284 @@
...
+
+static int DetectByteMathParseTest03(void)
+{
+    /* bytes value invalid */
+    DetectByteMathData *bmd = DetectByteMathParse("bytes 11, offset 2, oper +, "
+                                                  "rvalue 39, result bar", NULL);
+    if (bmd != NULL)
+        DetectByteMathFree(bmd);
+
+    FAIL_IF(bmd != NULL);
```
Comment (victorjulien) [inline src/detect-bytemath.c:814]: this should move up, the the `if (bmd != NULL)` condition can be removed.

General rule in (new) unittests: no conditions.

Comment (victorjulien) [reply src/detect-bytemath.c:814]: Also, tests should clean up memory on their success path. But can ignore failure path.

File: src/detect-bytemath.c [hunk 3]
```diff
@@ -0,0 +1,1284 @@
...
+
+    if (bmd != NULL)
+        DetectByteMathFree(bmd);
+
+    PASS;
+}
+
+static int MyUTHPacketMatchSig(Packet *p, const char *sig);
```
Comment (victorjulien) [inline src/detect-bytemath.c:1061]: not needed

File: src/detect-bytemath.c [hunk 4]
```diff
@@ -0,0 +1,1284 @@
...
+    FAIL_IF(PacketAlertCheck(p, de_ctx->sig_list->id) == 1);
+
+    Signature *s = de_ctx->sig_list;
+    FAIL_IF(s->sm_lists_tail[DETECT_SM_LIST_PMATCH] == NULL);
+
+    SigMatch *sm = NULL;
+    sm = s->sm_lists[DETECT_SM_LIST_PMATCH];
+    FAIL_IF(sm->type != DETECT_CONTENT);
+
+    SigGroupCleanup(de_ctx);
```
Comment (victorjulien) [inline src/detect-bytemath.c:1093]: not needed, `DetectEngineCtxFree` calls both `SigGroupCleanup` and `SigCleanSignatures`

**Comment by jlucovsky** [comment]
Continued in #4911

**Pull request closed without merging**